File: week1/이지훈/가장 먼 노드.java
import java.util.*;

public class LV3 {
    private static List<List<Integer>> graph = new ArrayList<>(); // 각 노드간의 경로 저장
    private static int[] dist; // 1번 노드 부터 각 노드들까지의 거리 저장
    private static int max; // 1번 노드로부터 가장 멀리 떨어진 노드와의 거리

    public int solution(int n, int[][] edge) {
        dist = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < edge.length; i++) {
            int x = edge[i][0];
            int y = edge[i][1];

            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        bfs(n);

        int result = 0;
        for (int num : dist) {
            if(num == max) result++;
        }

        return result;
    }

    private static void bfs(int n){
        Queue<Integer> q = new LinkedList<>();
        boolean[] visit = new boolean[n + 1];
        q.add(1);
        visit[1] = true;

        while (!q.isEmpty()) {
            Integer now = q.poll();

            for (int next : graph.get(now)) {
                if (!visit[next]) {
                    visit[next] = true;
                    q.add(next);
                    dist[next] = dist[now] + 1;
                    max = Math.max(max, dist[next]);
                }
            }
        }
    }
}

File: week1/한지희/가장 먼 노드.kt
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.max

class TheFarthestNode {
    private lateinit var isVisited: BooleanArray
    private lateinit var graph: Array<ArrayList<Int>>
    private lateinit var distance: IntArray

    fun solution(n: Int, edge: Array<IntArray>): Int {
        var answer = 0
        graph = Array(n + 1) { ArrayList() }
        isVisited = BooleanArray(n + 1)
        distance = IntArray(n + 1)

        edge.forEach { ints ->
            graph[ints[0]].add(ints[1])
            graph[ints[1]].add(ints[0])
        }

        bfs(1)

        var maxCount = 0
        for (i in 1..n) maxCount = max(maxCount, distance[i])
        for (i in 1..n) if (distance[i] == maxCount) answer++

        return answer
    }

    private fun bfs(idx: Int) {
        val queue: Queue<Int> = LinkedList()
        isVisited[idx] = true
        queue.offer(idx)

        while (queue.isNotEmpty()) {
            val now = queue.poll()
            for (i in graph[now].indices) {
                val next = graph[now][i]
                if (!isVisited[next]) {
                    distance[next] = distance[now] + 1
                    isVisited[next] = true
                    queue.offer(next)
                }
            }
        }
    }
}